src/axom/primal/operators/detail/winding_number_2d_memoization.hpp

```diff
@@ -0,0 +1,172 @@
+    m_numControlPoints = a_curve.getNumControlPoints();
+    m_numSpans = a_curve.getNumKnotSpans();
+
+    m_bezierSubdivisionMaps.resize(m_numSpans);
+    auto beziers = a_curve.extractBezier();
+
+    for(int idx = 0; idx < m_numSpans; ++idx)
+    {
+      m_bezierSubdivisionMaps[idx][std::make_pair(0, 0)] = SubdivisionData<T>(beziers[idx], false);
+    }
+
+    m_initPoint = a_curve[0];
+    m_endPoint = a_curve[m_numControlPoints - 1];
+  }
+
+  /// \brief Initialize the cache with the data for a single Bezier curve
+  NURBSCurveGWNCache(const BezierCurve<T, 2>& a_curve)
+  {
+    m_boundingBox = a_curve.boundingBox();
+    m_degree = a_curve.getOrder();
+    m_numControlPoints = a_curve.getOrder() + 1;
+
+    if(a_curve.getOrder() <= 0)
+    {
+      m_numSpans = 0;
+      m_bezierSubdivisionMaps.clear();
+
+      m_initPoint = m_endPoint = Point<T, 2> {0.0, 0.0};
+    }
+    else
+    {
+      m_numSpans = 1;
+      m_bezierSubdivisionMaps.resize(1);
+
+      m_initPoint = a_curve[0];
+      m_endPoint = a_curve[m_degree];
+
+      m_bezierSubdivisionMaps[0][std::make_pair(0, 0)] = SubdivisionData<T>(a_curve, false);
+    }
+  }
+
+  /// \brief Query the map. If curve is not found, add it and it's pair from subdivision
+  const SubdivisionData<T>& getSubdivisionData(int idx, int refinementLevel, int refinementIndex) const
+  {
+    auto hash_key = std::make_pair(refinementLevel, refinementIndex);
+
+    if(m_bezierSubdivisionMaps[idx].find(hash_key) == m_bezierSubdivisionMaps[idx].end())
+    {
+      const SubdivisionData<T>& supercurve_data =
+        m_bezierSubdivisionMaps[idx][std::make_pair(refinementLevel - 1, refinementIndex / 2)];
+
+      BezierCurve<T, 2> sub1, sub2;
+      supercurve_data.curve.split(0.5, sub1, sub2);
+
+      // Make keys for the requested curve and its "sibling" in the heirarchy
+      const auto key1 = std::make_pair(refinementLevel, refinementIndex - refinementIndex % 2);
+      const auto key2 = std::make_pair(refinementLevel, refinementIndex - refinementIndex % 2 + 1);
+
+      // Populate the map
+      m_bezierSubdivisionMaps[idx][key1] =
+        SubdivisionData<T>(sub1, supercurve_data.isConvexControlPolygon);
+      m_bezierSubdivisionMaps[idx][key2] =
+        SubdivisionData<T>(sub2, supercurve_data.isConvexControlPolygon);
+    }
+
+    return m_bezierSubdivisionMaps[idx][hash_key];
+  }
+
+  auto getNumKnotSpans() const { return m_numSpans; }
+  auto boundingBox() const { return m_boundingBox; }
+  auto getNumControlPoints() const { return m_numControlPoints; }
+  auto getDegree() const { return m_degree; }
+
+  auto getInitPoint() const { return m_initPoint; }
+  auto getEndPoint() const { return m_endPoint; }
+
+private:
+  BoundingBox<T, 2> m_boundingBox;
+  int m_numControlPoints;
+  int m_degree;
+  int m_numSpans;
+
+  Point<T, 2> m_initPoint, m_endPoint;
+
+  mutable axom::Array<std::map<std::pair<int, int>, SubdivisionData<T>>> m_bezierSubdivisionMaps;
```

Comment:
Possible optimization for the future -- this might be more efficient w/ a hashmap (perhaps `axom::FlatMap` ?) than w/ a `std::map`.

(no need to make this change or even test it out in this PR)

Reply:
Right, I remember we talked about how `std::map` uses trees, right? It feels like it would be a natural fit for a subdivision structure like this, but I suppose there's no guarantee that map[3, 2], for example, would put the data at the 2nd index in the 3rd level (or, for that matter, if that'd even be the most efficient way to do it).